src/main/scala/mesosphere/marathon/tasks/TaskQueue.scala

```diff
@@ -39,33 +50,56 @@ class TaskQueue {
     * @param app The app
     * @return count
     */
-  def count(app: AppDefinition): Int = queue.asScala.count(_.app.id == app.id)
+  def count(app: AppDefinition): Int = apps.get((app.id, app.version)).map(_.count.get()).getOrElse(0)
 
   def purge(appId: PathId): Unit = {
-    val retained = queue.asScala.filterNot(_.app.id == appId)
-    removeAll()
-    queue.addAll(retained.asJavaCollection)
+    for {
+      QueuedTask(app, _, _) <- apps.values
+      if app.id == appId
+    } apps.remove(app.id -> app.version)
   }
 
   /**
     * Retains only elements that satisfy the supplied predicate.
     */
   def retain(f: (QueuedTask => Boolean)): Unit =
-    queue.iterator.asScala.foreach { qt => if (!f(qt)) queue.remove(qt) }
-
-  def addAll(xs: Seq[QueuedTask]): Unit = queue.addAll(xs.asJavaCollection)
-
-  def removeAll(): Seq[QueuedTask] = {
-    val builder = new java.util.ArrayList[QueuedTask]()
-    queue.drainTo(builder)
-    builder.asScala.to[Seq]
+    apps.values.foreach {
+      case qt @ QueuedTask(app, _, _) => if (!f(qt)) apps.remove(app.id -> app.version)
+    }
+
+  def pollMatching[B](f: AppDefinition => Option[B]): Option[B] = {
```

Comment:
Unfortunately a PF is not possible here, because we use it for complex things (task to offer matching). I also don't really like the name, any ideas for a better one?